### ISSUE TITLE
Cannot pass Optional to function expecting generic type argument

### ISSUE THREAD

**Issue by seh9000** [microsoft/pyright#684]
**Describe the bug**
A type error is raised when an 'Optional' type is passed to a generic function which expects a generic type as an argument. 

**To Reproduce**
Invoke pyright on the below code.

**Expected behavior**
I expect the type T to be the optional and for type checking to pass. It appears that pyright is splitting the Union which represents the optional and type checking the Value and None against the function which expects a type argument. I would expect the t.Type[T] to resolve the T as the Union not apply them separately.

**Screenshots or Code**
```
import typing as t

T = t.TypeVar('T')
def func(o: t.Type[T]) -> T:
    ...

x = t.Optional[int]
func(x)
```

In this example I expect T to resolve to t.Optional[int] not int and None separately.

**VS Code extension or command-line**
Command line using git commit hash 74e375dccb90b365b6c0db45ad0f8c5b016bf4bd

**Additional context**
None

**Comment by erictraut** [comment]
`Optional[T]` isn't a type. It's a shorthand for `Union[T, None]`, and `None` is not a type, it's an instance of `NoneType`. If you want your function to accept either a type T or None, you should declare it as:

```
def func(o: t.Optional[t.Type[T]]) -> T:
    ...
```
or you can declare a type alias
```
OptionalType = t.Optional[t.Type[T]]
def func(o: OptionalType[T]) -> T:
    ...
```

**Comment by seh9000** [comment]
How can I then accept something generically which may sometimes take the optional and sometimes not? Should I create an overload for it?

**Comment by erictraut** [comment]
Yes, an overload would work.

**Comment by seh9000** [comment]
Sorry I edited my comment late, I'm still confused as to how a Union would not be a type. It seems generically not useful to treat it that way.

**Comment by erictraut** [comment]
Sorry if my response was confusing. `Union[T, None]` is a type, but it's not a `Type`. In other words, it's not type that should ever match against `Type[T]`.

When a function specifies that one of its parameters must be of Type[T], there is an expectation that the parameter is a class that can be instantiated. You cannot instantiate a Union or a None.
